### PR TITLE
✨ Track number of crossings

### PR DESCRIPTION
**Pull request by simon1hofmann**
## Description

Add a tracker for the number of crossings in a layout that is automatically updated when a crossing is added or removed.

## Checklist:

- [ ] The pull request only contains commits that are related to it.
- [ ] I have added appropriate tests and documentation.
- [ ] I have added a changelog entry.
- [ ] I have created/adjusted the Python bindings for any new or updated functionality.
- [ ] I have made sure that all CI jobs on GitHub pass.
- [ ] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: cli/stores.hpp [hunk 1]
```diff
@@ -242,6 +243,7 @@ ALICE_LOG_STORE_STATISTICS(fiction::gate_layout_t, layout)
             {"outputs", lyt_ptr->num_pos()},
             {"gates", lyt_ptr->num_gates()},
             {"wires", lyt_ptr->num_wires()},
+            {"crossings", lyt_ptr->num_crossings()},
             {"layout", {{"x-size", lyt_ptr->x() + 1}, {"y-size", lyt_ptr->y() + 1}, {"area", lyt_ptr->area()}}},
             // {"free tiles", area - (gate_tiles + wire_tiles - crossings)},  // free tiles in ground layer
             // {"crossings", crossings},
```
Comment (marcelwa) [inline cli/stores.hpp:249]:
```suggestion
```
This may go now :smile:

File: test/layouts/gate_level_layout.cpp [hunk 2]
```diff
@@ -1073,6 +1079,7 @@ TEST_CASE("Move nodes", "[gate-level-layout]")
```
Comment (marcelwa) [inline test/layouts/gate_level_layout.cpp:1079]: Could you add a test particularly moving around crossings and check if their tracking behaves correctly?

**Review by marcelwa — CHANGES_REQUESTED** [review]

**Review by marcelwa — COMMENTED** [review]
Only now thought of the possibility to also log the crossing numbers in the stats of our various physical design algorithms. Maybe that would be nice. What do you think?

**Comment by simon1hofmann** [comment]
> Only now thought of the possibility to also log the crossing numbers in the stats of our various physical design algorithms. Maybe that would be nice. What do you think?

Good idea 👍

**Comment by marcelwa** [comment]
> > Only now thought of the possibility to also log the crossing numbers in the stats of our various physical design algorithms. Maybe that would be nice. What do you think?
> 
> Good idea 👍

Many thanks! The adjusted `_stats` classes also need to be reflected in the Python bindings, though.

**Comment by marcelwa** [comment]
> > > > Only now thought of the possibility to also log the crossing numbers in the stats of our various physical design algorithms. Maybe that would be nice. What do you think?
> > > 
> > > 
> > > Good idea 👍
> > 
> > 
> > Many thanks! The adjusted `_stats` classes also need to be reflected in the Python bindings, though.
> 
> They are reflected in the python bindings that make use of members of the `_stats` object. Is this what you meant?

I see that we have some inconsistencies going on in how the `_stats` classes are exposed to Python. Sometimes, we use

```python
    py::class_<fiction::exact_physical_design_stats>(m, "exact_stats", DOC(fiction_exact_physical_design_stats))
        .def(py::init<>())
        .def("__repr__",
             [](const fiction::exact_physical_design_stats& stats)
             {
                 std::stringstream stream{};
                 stats.report(stream);
                 return stream.str();
             })

        ;
```
(no access to the attributes, just printing via `__repr__` function)

sometimes, we do

```python
py::class_<fiction::technology_mapping_stats>(m, "technology_mapping_stats", DOC(fiction_technology_mapping_stats))
        .def(py::init<>())

        .def("report", &fiction::technology_mapping_stats::report, DOC(fiction_technology_mapping_stats_report))

        ;
```
(no access to the attributes, just printing via a `report` function)

and sometimes we do

```python
py::class_<fiction::graph_oriented_layout_design_stats>(m, "graph_oriented_layout_design_stats",
                                                            DOC(fiction_graph_oriented_layout_design_stats))
        .def(py::init<>())
        .def("__repr__",
             [](const fiction::graph_oriented_layout_design_stats& stats)
             {
                 std::stringstream stream{};
                 stats.report(stream);
                 return stream.str();
             })
        .def_readonly("time_total", &fiction::graph_oriented_layout_design_stats::time_total,
                      DOC(fiction_graph_oriented_layout_design_stats_duration))
        .def_readonly("x_size", &fiction::graph_oriented_layout_design_stats::x_size,
                      DOC(fiction_graph_oriented_layout_design_stats_x_size))
        .def_readonly("y_size", &fiction::graph_oriented_layout_design_stats::y_size,
                      DOC(fiction_graph_oriented_layout_design_stats_y_size))
        .def_readonly("num_gates", &fiction::graph_oriented_layout_design_stats::num_gates,
                      DOC(fiction_graph_oriented_layout_design_stats_num_gates))
        .def_readonly("num_wires", &fiction::graph_oriented_layout_design_stats::num_wires,
                      DOC(fiction_graph_oriented_layout_design_stats_num_wires))

        ;
```
(read-only access to the attributes and printing via `__repr__`)

In my opinion, the last one is the best. However, we could also add the `report` function there, too, for completeness. And then adjust all other `_stats` in pyfiction to follow this best practice.

File: bindings/pyfiction/include/pyfiction/algorithms/verification/equivalence_checking.hpp [hunk 3]
```diff
@@ -60,8 +60,21 @@ inline void equivalence_checking(pybind11::module& m)
...
         .def(py::init<>())
+        .def_readwrite("eq", &fiction::equivalence_checking_stats::eq, DOC(fiction_equivalence_checking_stats_eq))
+        .def_readwrite("tp_spec", &fiction::equivalence_checking_stats::tp_spec,
+                       DOC(fiction_equivalence_checking_stats_tp_spec))
+        .def_readwrite("tp_impl", &fiction::equivalence_checking_stats::tp_impl,
+                       DOC(fiction_equivalence_checking_stats_tp_impl))
+        .def_readwrite("tp_diff", &fiction::equivalence_checking_stats::tp_diff,
+                       DOC(fiction_equivalence_checking_stats_tp_diff))
         .def_readwrite("counter_example", &fiction::equivalence_checking_stats::counter_example,
                        DOC(fiction_equivalence_checking_stats_counter_example))
```
Comment (marcelwa) [inline bindings/pyfiction/include/pyfiction/algorithms/verification/equivalence_checking.hpp:71]: They can all be instantiated `def_readonly`

File: bindings/pyfiction/include/pyfiction/algorithms/simulation/sidb/time_to_solution.hpp [hunk 4]
```diff
@@ -52,6 +52,14 @@ inline void time_to_solution(pybind11::module& m)
...
         .def(py::init<>())
+        .def("__repr__",
+             [](const fiction::time_to_solution_stats& stats)
+             {
+                 std::stringstream stream{};
+                 stats.report(stream);
+                 return stream.str();
+             })
+        .def("report", &fiction::time_to_solution_stats::report, DOC(fiction_time_to_solution_stats_report))
         .def_readwrite("time_to_solution", &fiction::time_to_solution_stats::time_to_solution,
```
Comment (marcelwa) [inline bindings/pyfiction/include/pyfiction/algorithms/simulation/sidb/time_to_solution.hpp:63]: These are still `def_readwrite`

File: bindings/pyfiction/include/pyfiction/algorithms/simulation/sidb/critical_temperature.hpp [hunk 5]
```diff
@@ -57,6 +57,14 @@ inline void critical_temperature(pybind11::module& m)
...
         .def(py::init<>())
+        .def("__repr__",
+             [](const fiction::critical_temperature_stats& stats)
+             {
+                 std::stringstream stream{};
+                 stats.report(stream);
+                 return stream.str();
+             })
+        .def("report", &fiction::critical_temperature_stats::report, DOC(fiction_critical_temperature_stats_report))
         .def_readwrite("algorithm_name", &fiction::critical_temperature_stats::algorithm_name,
```
Comment (marcelwa) [inline bindings/pyfiction/include/pyfiction/algorithms/simulation/sidb/critical_temperature.hpp:68]: These are still `def_readwrite` but can be made `def_readonly`

**Review by marcelwa — CHANGES_REQUESTED** [review]

File: bindings/pyfiction/include/pyfiction/algorithms/verification/equivalence_checking.hpp [hunk 6]
```diff
@@ -60,8 +60,21 @@ inline void equivalence_checking(pybind11::module& m)
...
                                                     DOC(fiction_equivalence_checking_stats))
         .def(py::init<>())
+        .def_readonly("eq", &fiction::equivalence_checking_stats::eq, DOC(fiction_equivalence_checking_stats_eq))
+        .def_readonly("tp_spec", &fiction::equivalence_checking_stats::tp_spec,
+                      DOC(fiction_equivalence_checking_stats_tp_spec))
+        .def_readonly("tp_impl", &fiction::equivalence_checking_stats::tp_impl,
+                      DOC(fiction_equivalence_checking_stats_tp_impl))
+        .def_readonly("tp_diff", &fiction::equivalence_checking_stats::tp_diff,
+                      DOC(fiction_equivalence_checking_stats_tp_diff))
         .def_readwrite("counter_example", &fiction::equivalence_checking_stats::counter_example,
```
Comment (simon1hofmann) [inline bindings/pyfiction/include/pyfiction/algorithms/verification/equivalence_checking.hpp:70]: @marcelwa What about `counter_example`?

Comment (marcelwa) [reply bindings/pyfiction/include/pyfiction/algorithms/verification/equivalence_checking.hpp:70]: Let's make it `def_readonly` too for consistency

**Review by marcelwa — APPROVED** [review]
Many thanks for the addition and also for dealing with the unexpected changes :pray:

**Pull request merged**